File: tests/unit/bkl_power_test.c
/* SPDX-License-Identifier: MIT */
/* The real power control flow; hardware ports/IDT and VFS admission are leaves.
 * Actual drain semantics are covered by the VFS mount concurrency gate. */
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <pthread.h>
#include <sched.h>
#include "power.h"
#include "vfs.h"
#include "acpi.h"
static unsigned begins,ends,active,overlap,syncs,writes,unsafe_write;static int mode;
int vfs_drain_begin(struct vfs_drain *t){
    __atomic_add_fetch(&begins,1,__ATOMIC_SEQ_CST);if(mode==2)return -16;
    t->active=1;if(__atomic_fetch_add(&active,1,__ATOMIC_SEQ_CST))__atomic_add_fetch(&overlap,1,__ATOMIC_SEQ_CST);return 0;
}
void vfs_drain_end(struct vfs_drain *t){if(t->active){t->active=0;__atomic_sub_fetch(&active,1,__ATOMIC_SEQ_CST);__atomic_add_fetch(&ends,1,__ATOMIC_SEQ_CST);}}
int logitfs_sync(void){
    __atomic_add_fetch(&syncs,1,__ATOMIC_SEQ_CST);
#ifdef POWER_CTL_RACE
    while(__atomic_load_n(&syncs,__ATOMIC_SEQ_CST)<2)sched_yield();
#endif
    return mode==0?-1:0;
}
void kprintf(const char *fmt,...){(void)fmt;}
uint64_t time_mono_ns(void){static uint64_t t;return __atomic_add_fetch(&t,1000000,__ATOMIC_RELAXED);}
int acpi_pm1_cnt(uint32_t *a,uint32_t *b){*a=0x604;*b=0;return 0;}
int acpi_reset_reg(struct acpi_gas *r,uint8_t *v){(void)r;(void)v;return -1;}
static void write_port(void){
    if(__atomic_load_n(&active,__ATOMIC_SEQ_CST)!=1||!__atomic_load_n(&syncs,__ATOMIC_SEQ_CST))__atomic_add_fetch(&unsafe_write,1,__ATOMIC_SEQ_CST);
    __atomic_add_fetch(&writes,1,__ATOMIC_SEQ_CST);
}
void outb(uint16_t p,uint8_t v){(void)p;(void)v;write_port();}
void outw(uint16_t p,uint16_t v){(void)p;(void)v;write_port();}
static void *power(void *arg){return (void *)(intptr_t)((intptr_t)arg?kernel_reboot():kernel_poweroff());}
int main(int argc,char **argv){
    mode=argc>1&&!strcmp(argv[1],"hardware")?1:argc>1&&!strcmp(argv[1],"begin")?2:0;
    pthread_t a,b;pthread_create(&a,0,power,0);pthread_create(&b,0,power,0);void *ra,*rb;pthread_join(a,&ra);pthread_join(b,&rb);
    int bad=0;
    if(overlap){puts("FAIL: power requests have one transition owner");bad++;}
    if(active||(mode!=2&&ends!=2)||((intptr_t)ra>=0)||((intptr_t)rb>=0)){puts("FAIL: failed power action restores VFS admission");bad++;}
    if(unsafe_write||(mode!=1&&writes)){puts("FAIL: hardware operation follows drain and successful sync");bad++;}
    if(mode==1){if(kernel_reboot()>=0||active||ends!=3){puts("FAIL: returned reset attempt restores admission");bad++;}}
    printf("BKL_POWER begins=%u ends=%u active=%u failures=%d\n",begins,ends,active,bad);return bad?1:0;
}

File: tests/unit/bkl_power_test.py
#!/usr/bin/env python3
# SPDX-License-Identifier: MIT
import argparse,os,pathlib,re,subprocess
p=argparse.ArgumentParser();p.add_argument("--build",default="/tmp/logitos-bkl-power");p.add_argument("--source-dir");a=p.parse_args()
r=pathlib.Path(__file__).resolve().parents[2];b=pathlib.Path(a.build);b.mkdir(parents=True,exist_ok=True)
src=pathlib.Path(a.source_dir) if a.source_dir else r/"c/kernel/core"
(b/"power.h").write_text((src/"power.h").read_text())
(b/"wait.h").write_text("#include <pthread.h>\nstruct mutex {pthread_mutex_t m;};\n#define MUTEX_INIT {PTHREAD_MUTEX_INITIALIZER}\nstatic inline void mutex_lock(struct mutex *m){pthread_mutex_lock(&m->m);}\nstatic inline void mutex_unlock(struct mutex *m){pthread_mutex_unlock(&m->m);}\n")
(b/"io.h").write_text("#include <stdint.h>\nvoid outb(uint16_t,uint8_t);\nvoid outw(uint16_t,uint16_t);\n")
s=(src/"power.c").read_text()
# A host cannot reset its IDT or run x86 pause on arm64. Only the privileged
# leaf statements are replaced; preparation/cancellation/port order is real.
s,n=re.subn(r'__asm__ volatile\s*\([^;]*\);','(void)0;',s)
if n!=5:raise SystemExit('power privileged leaf count changed: '+str(n))
base=[os.environ.get("CC","clang"),"-std=c11","-O1","-g","-pthread","-fsanitize=address,undefined","-I"+str(b)]
for d in ("c/kernel/core","c/kernel/cpu","c/kernel/cpu/acpi","c/kernel/cpu/irq","c/kernel/cpu/smp","c/kernel/cpu/acpi","c/kernel/cpu/irq","c/kernel/cpu/smp","c/fs","c/fs/vfs","c/fs/logitfs","c/fs/cache","c/fs/ramfs","c/fs/ctl","c/fs/procfs","c/fs/vfs","c/fs/logitfs","c/fs/cache","c/fs/ramfs","c/fs/ctl","c/fs/procfs","c/drivers/block","include/abi"):base += ["-I"+str(r/d)]
def run(name,code,mode,defs=(),expect=None):
    source=b/(name+".c");source.write_text(code);exe=b/name
    subprocess.run(base+list(defs)+[str(r/"tests/unit/bkl_power_test.c"),str(source),"-o",str(exe)],check=True,cwd=r)
    q=subprocess.run([str(exe),mode],capture_output=True,text=True,timeout=30);out=q.stdout+q.stderr;(b/(name+".log")).write_text(out)
    if expect:
        if q.returncode==0 or expect not in out:raise SystemExit("control missed intended assertion: "+name+"\n"+out)
        print("CONTROL",name,"caught:",expect)
    elif q.returncode:raise SystemExit(out)
    else:print(out,end="")
none=s.replace('mutex_lock(&power_lock);','(void)power_lock;').replace('mutex_unlock(&power_lock);','(void)power_lock;')
run("power-no-owner",none,"sync",["-DPOWER_CTL_RACE"],"FAIL: power requests have one transition owner")
leak=s.replace('vfs_drain_end(token);','(void)token;')
run("power-no-rollback",leak,"sync",expect="FAIL: failed power action restores VFS admission")
for mode in ("sync","begin","hardware"):run("power-"+mode+"-positive",s,mode)
